monkey/infection_monkey/master/automated_master.py

```diff
@@ -0,0 +1,152 @@
+import logging
+import threading
+import time
+from typing import Dict, List
+
+from infection_monkey.i_control_channel import IControlChannel
+from infection_monkey.i_master import IMaster
+from infection_monkey.i_puppet import IPuppet
+from infection_monkey.telemetry.messengers.i_telemetry_messenger import ITelemetryMessenger
+from infection_monkey.telemetry.system_info_telem import SystemInfoTelem
+from infection_monkey.utils.timer import Timer
+
+CHECK_ISLAND_FOR_STOP_COMMAND_INTERVAL_SEC = 5
+CHECK_FOR_TERMINATE_INTERVAL_SEC = CHECK_ISLAND_FOR_STOP_COMMAND_INTERVAL_SEC / 5
+SHUTDOWN_TIMEOUT = 5
+
+logger = logging.getLogger()
+
+
+class AutomatedMaster(IMaster):
+    def __init__(
+        self,
+        puppet: IPuppet,
+        telemetry_messenger: ITelemetryMessenger,
+        control_channel: IControlChannel,
+    ):
+        self._puppet = puppet
+        self._telemetry_messenger = telemetry_messenger
+        self._control_channel = control_channel
+
+        self._stop = threading.Event()
+        self._master_thread = threading.Thread(target=self._run_master_thread, daemon=True)
+        self._simulation_thread = threading.Thread(target=self._run_simulation, daemon=True)
+
+    def start(self):
+        logger.info("Starting automated breach and attack simulation")
+        self._master_thread.start()
+        self._master_thread.join()
+        logger.info("The simulation has been shutdown.")
+
+    def terminate(self):
+        logger.info("Stopping automated breach and attack simulation")
+        self._stop.set()
+
+        if self._master_thread.is_alive():
+            self._master_thread.join()
+
+    def _run_master_thread(self):
+        self._simulation_thread.start()
+
+        self._wait_for_master_stop_condition()
+
+        logger.debug("Waiting for the simulation thread to stop")
+        self._simulation_thread.join(SHUTDOWN_TIMEOUT)
+
+        if self._simulation_thread.is_alive():
+            logger.warn("Timed out waiting for the simulation to stop")
+            # Since the master thread and all child threads are daemon threads, they will be
+            # forcefully killed when the program exits.
+            logger.warn("Forcefully killing the simulation")
+
+    def _wait_for_master_stop_condition(self):
+        timer = Timer()
+        timer.set(CHECK_ISLAND_FOR_STOP_COMMAND_INTERVAL_SEC)
+
+        while self._master_thread_should_run():
+            if timer.is_expired():
+                self._check_for_stop()
+                timer.reset()
+
+            time.sleep(CHECK_FOR_TERMINATE_INTERVAL_SEC)
+
+    def _check_for_stop(self):
+        if self._control_channel.should_agent_stop():
+            logger.debug('Received the "stop" signal from the Island')
+            self._stop.set()
+
+    def _master_thread_should_run(self):
+        return (not self._stop.is_set()) and self._simulation_thread.is_alive()
```

Comment:
So the Agent checks whether the simulation is still running every second, and checks whether it should still be running as per the Island's instructions every 5 seconds?

Reply:
I've tried to make it more readable, not sure it's better:
```
    def _wait_for_master_stop_condition2(self):
        self.timer.set(CHECK_REMOTE_TERMINATION_INTERVAL_SEC)

        while self._check_for_master_stop():
            time.sleep(CHECK_FOR_TERMINATE_INTERVAL_SEC)
            
    def _check_for_master_stop(self):
        return self._master_thread_should_run() or self._check_for_remote_termination()
    
    def _check_for_remote_termination(self):
        if self.timer.is_expired():
            if self._control_channel.should_agent_stop():
                logger.debug('Received the "stop" signal from the Island')
                return True
            self.timer.reset()
            return False
        else:
            return False
```

Reply:
I'll stick with the original for now. The suggested code has a function call `_check_for_remote_termination()` that may or may not check for termination.